ssm_jax/nlgssm/extended_inference.py

```diff
@@ -0,0 +1,165 @@
+            filtered_means: (T,D_hid) array,
+                E[x_t | y_{1:t}, u_{1:t}].
+            filtered_covariances: (T,D_hid,D_hid) array,
+                Cov[x_t | y_{1:t}, u_{1:t}].
+            smoothed_means: (T,D_hid) array,
+                E[x_t | y_{1:T}, u_{1:T}].
+            smoothed_covs: (T,D_hid,D_hid) array of smoothed marginal covariances,
+                Cov[x_t | y_{1:T}, u_{1:T}].
+            smoothed_cross: (T-1, D_hid, D_hid) array of smoothed cross products,
+                E[x_t x_{t+1}^T | y_{1:T}, u_{1:T}].
+    """
+    marginal_loglik: chex.Scalar = None
+    filtered_means: chex.Array = None
+    filtered_covariances: chex.Array = None
+    smoothed_means: chex.Array = None
+    smoothed_covariances: chex.Array = None
+    smoothed_cross_covariances: chex.Array = None
+
+
+# Helper functions
+_get_params = lambda x, dim, t: x[t] if x.ndim == dim+1 else x
+
+def _predict(m, S, f, F, Q, u):
+    """Predict next mean and covariance using first-order additive EKF
+
+        p(x_{t+1}) = \int N(x_t | m, S) N(x_{t+1} | f(x_t, u), Q)
+                    = N(x_{t+1} | f(m, u), F(m, u) S F(m, u)^T + Q)
+
+    Args:
+        m (D_hid,): prior mean.
+        S (D_hid,D_hid): prior covariance.
+        f (Callable): dynamics function.
+        F (Callable): Jacobian of dynamics function.
+        Q (D_hid,D_hid): dynamics covariance matrix.
+        u (D_in,): inputs.
+
+    Returns:
+        mu_pred (D_hid,): predicted mean.
+        Sigma_pred (D_hid,D_hid): predicted covariance.
+    """
+    F_x = F(m, u)
+    mu_pred = f(m, u)
+    Sigma_pred = F_x @ S @ F_x.T + Q
+    return mu_pred, Sigma_pred
+
+
+def _condition_on(m, S, h, H, R, u, y):
+    """Condition a Gaussian potential on a new observation
+      p(x_t | y_t, u_t, y_{1:t-1}, u_{1:t-1})
+        propto p(x_t | y_{1:t-1}, u_{1:t-1}) p(y_t | x_t, u_t)
+        = N(x_t | m, S) N(y_t | h_t(x_t, u_t), R_t)
+        = N(x_t | mm, SS)
+    where
+        mm = m + K*(y - yhat) = mu_cond
+        yhat = h(m, u)
+        K = S * H(m, u)' * (R + H(m, u) * S * H(m, u)')^{-1}
+        L = I - K*H(m, u)
+        SS = L * S * L' + K * R * K' = Sigma_cond
+    **Note! This can be done more efficiently when R is diagonal.**
+
+   Args:
+        m (D_hid,): prior mean.
+        S (D_hid,D_hid): prior covariance.
+        h (Callable): emission function.
+        H (Callable): Jacobian of emission function.
+        R (D_obs,D_obs): emission covariance matrix.
+        u (D_in,): inputs.
+        y (D_obs,): observation.
+
+    Returns:
+        mu_cond (D_hid,): filtered mean.
+        Sigma_cond (D_hid,D_hid): filtered covariance.
+    """
+    H_x = H(m, u)
+    K = jnp.linalg.solve(R + H_x @ S @ H_x.T, H_x @ S).T
+    ImKH = jnp.eye(m.shape[-1]) - K @ H_x
+    Sigma_cond = ImKH @ S @ ImKH.T + K @ R @ K.T
+    mu_cond = m + K @ (y - h(m, u))
+    return mu_cond, Sigma_cond
+
+
+def essm_filter(params, emissions, inputs=None):
```

Comment:
Should we call this `ekf_filter` or `extended_kalman_filter`?
(I know https://github.com/probml/ssm-jax/blob/main/ssm_jax/lgssm/inference.py#L114 is called `lgssm_filter` but I am not wild about that...)

Reply:
I like `extended_kalman_filter`!

Reply:
+1 for `extended_kalman_filter`.

Happy to revisit the discussion about `lgssm_filter` vs `kalman_fitler`. The latter is certainly more standard.